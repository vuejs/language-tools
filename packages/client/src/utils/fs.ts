import * as vscode from 'vscode';

export async function exists(uri: vscode.Uri) {
    try {
        await vscode.workspace.fs.stat(uri);
        return true;
    }
    catch {
        return false;
    }
}
