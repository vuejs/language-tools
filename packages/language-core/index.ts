export * from './lib/codegen/template';
export * from './lib/languageModule';
export * from './lib/parsers/scriptSetupRanges';
export * from './lib/plugins';
export * from './lib/virtualFile/vueFile';
export * from './lib/types';
export * from './lib/utils/ts';
export * from './lib/utils/parseSfc';

export * as scriptRanges from './lib/parsers/scriptRanges';
export * from './lib/utils/shared';
export { tsCodegen } from './lib/plugins/vue-tsx';

export * from '@volar/language-core';
export type * as CompilerDOM from '@vue/compiler-dom';
