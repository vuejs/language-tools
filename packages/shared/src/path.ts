import { URI } from 'vscode-uri';
import * as path from 'typesafe-path';

export function getPathOfUri(uri: string) {
	const _uri = URI.parse(uri);
	if (_uri.scheme === 'file') {
		return _uri.fsPath.replace(/\\/g, '/') as path.PosixPath;
	}
	else {
		return '/__uri__/' + uri.replace('://', '__uriScheme__/') as path.PosixPath;
	}
}

export function normalizeFileName(fsPath: string) {
	return URI.file(fsPath).fsPath.replace(/\\/g, '/') as path.PosixPath;
}

export function getUriByPath(path: string) {
	if (path.startsWith('/__uri__/')) {
		return path.replace('/__uri__/', '').replace('__uriScheme__/', '://');
	}
	return URI.file(path).toString();
}

export function isFileInDir(fileName: path.OsPath, dir: path.OsPath) {
	const relative = path.relative(dir, fileName);
	return !!relative && !relative.startsWith('..') && !path.isAbsolute(relative);
}
