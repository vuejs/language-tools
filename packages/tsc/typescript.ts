import { run } from '.'

export deafult run(require.resolve('typescript/lib/typescript'))
