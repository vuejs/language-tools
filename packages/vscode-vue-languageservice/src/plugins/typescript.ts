import { TextDocument } from 'vscode-languageserver-textdocument';
import { definePlugin } from '../utils/definePlugin';
import type * as ts from 'typescript/lib/tsserverlibrary';
import * as ts2 from 'vscode-typescript-languageservice';

export { getSemanticTokenLegend, getTriggerCharacters } from 'vscode-typescript-languageservice';

export default definePlugin((host: {
    getTsLs: () => ts2.LanguageService,
    baseCompletionOptions?: ts.GetCompletionsAtPositionOptions,
}) => {

    return {

        doValidation(document, options) {
            if (isTsDocument(document)) {
                return host.getTsLs().doValidation(document.uri, options);
            }
        },

        doComplete(document, position, context) {
            if (isTsDocument(document)) {
                const options: ts.GetCompletionsAtPositionOptions = {
                    ...host.baseCompletionOptions,
                    triggerCharacter: context?.triggerCharacter as ts.CompletionsTriggerCharacter,
                    triggerKind: context?.triggerKind,
                };

                return host.getTsLs().doComplete(document.uri, position, options);
            }
        },

        doCompleteResolve(item) {
            return host.getTsLs().doCompletionResolve(item);
        },

        doHover(document, position) {
            if (isTsDocument(document)) {
                return host.getTsLs().doHover(document.uri, position);
            }
        },

        findDefinition(document, position) {
            if (isTsDocument(document)) {
                return host.getTsLs().findDefinition(document.uri, position);
            }
        },

        findTypeDefinition(document, position) {
            if (isTsDocument(document)) {
                return host.getTsLs().findTypeDefinition(document.uri, position);
            }
        },

        findImplementations(document, position) {
            if (isTsDocument(document)) {
                return host.getTsLs().findImplementations(document.uri, position);
            }
        },

        findReferences(document, position) {
            if (isTsDocument(document) || isJsonDocument(document)) {
                return host.getTsLs().findReferences(document.uri, position);
            }
        },

        findDocumentHighlights(document, position) {
            if (isTsDocument(document)) {
                return host.getTsLs().findDocumentHighlights(document.uri, position);
            }
        },

        findDocumentSymbols(document) {
            if (isTsDocument(document)) {
                return host.getTsLs().findDocumentSymbols(document.uri);
            }
        },

        findDocumentSemanticTokens(document, range, cancleToken) {
            if (isTsDocument(document)) {
                return host.getTsLs().getDocumentSemanticTokens(document.uri, range, cancleToken);
            }
        },

        findWorkspaceSymbols(query) {
            return host.getTsLs().findWorkspaceSymbols(query);
        },

        doCodeActions(document, range, context) {
            if (isTsDocument(document)) {
                return host.getTsLs().getCodeActions(document.uri, range, context);
            }
        },

        doCodeActionResolve(codeAction) {
            return host.getTsLs().doCodeActionResolve(codeAction);
        },

        doRenamePrepare(document, position) {
            if (isTsDocument(document)) {
                return host.getTsLs().prepareRename(document.uri, position);
            }
        },

        doRename(document, position, newName) {
            if (isTsDocument(document) || isJsonDocument(document)) {
                return host.getTsLs().doRename(document.uri, position, newName);
            }
        },

        doFileRename(oldUri, newUri) {
            return host.getTsLs().getEditsForFileRename(oldUri, newUri);
        },

        getFoldingRanges(document) {
            if (isTsDocument(document)) {
                return host.getTsLs().getFoldingRanges(document.uri);
            }
        },

        getSelectionRanges(document, positions) {
            if (isTsDocument(document)) {
                return host.getTsLs().getSelectionRanges(document.uri, positions);
            }
        },

        getSignatureHelp(document, position, context) {
            if (isTsDocument(document)) {
                return host.getTsLs().getSignatureHelp(document.uri, position, context);
            }
        },

        format(document, range, options) {
            if (isTsDocument(document)) {
                return host.getTsLs().doFormatting(document.uri, options, range);
            }
        },

        callHierarchy: {

            doPrepare(document, position) {
                if (isTsDocument(document)) {
                    return host.getTsLs().callHierarchy.doPrepare(document.uri, position);
                }
            },

            getIncomingCalls(item) {
                return host.getTsLs().callHierarchy.getIncomingCalls(item);
            },

            getOutgoingCalls(item) {
                return host.getTsLs().callHierarchy.getOutgoingCalls(item);
            },
        },
    };
});

export function isTsDocument(document: TextDocument) {
    return document.languageId === 'javascript' ||
        document.languageId === 'typescript' ||
        document.languageId === 'javascriptreact' ||
        document.languageId === 'typescriptreact';
}

export function isJsonDocument(document: TextDocument) {
    return document.languageId === 'json' ||
        document.languageId === 'jsonc';
}
