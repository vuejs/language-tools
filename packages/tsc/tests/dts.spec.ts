import * as path from 'path';
import * as fs from 'fs';
import * as ts from 'typescript';
import { describe, expect, it } from 'vitest';
import { proxyCreateProgram } from '@volar/typescript';
import * as vue from '@vue/language-core';
import { removeEmitGlobalTypes } from '..';

const workspace = path.resolve(__dirname, '../../../test-workspace/component-meta');
const normalizePath = (filename: string) => filename.replace(/\\/g, '/');
const normalizeNewline = (text: string) => text.replace(/\r\n/g, '\n');
const windowsPathReg = /\\/g;

describe('vue-tsc-dts', () => {
	const compilerOptions: ts.CompilerOptions = {
		rootDir: workspace,
		declaration: true,
		emitDeclarationOnly: true,
		allowNonTsExtensions: true,
	};
	const host = ts.createCompilerHost(compilerOptions);
	const options: ts.CreateProgramOptions = {
		host,
		rootNames: readFilesRecursive(workspace),
		options: compilerOptions
	};

	let vueOptions: vue.VueCompilerOptions;
	const createProgram = proxyCreateProgram(ts, ts.createProgram, (ts, options) => {
		const { configFilePath } = options.options;
		vueOptions = typeof configFilePath === 'string'
			? vue.createParsedCommandLine(ts, ts.sys, configFilePath.replace(windowsPathReg, '/')).vueOptions
			: vue.resolveVueCompilerOptions({ extensions: ['.vue', '.cext'] });
		const vueLanguagePlugin = vue.createVueLanguagePlugin(
			ts,
			id => id,
			options.host?.useCaseSensitiveFileNames?.() ?? false,
			() => '',
			() => options.rootNames.map(rootName => rootName.replace(windowsPathReg, '/')),
			options.options,
			vueOptions,
		);
		return [vueLanguagePlugin];
	});
	const program = createProgram(options);

	for (const intputFile of options.rootNames) {

		const expectedOutputFile = intputFile.endsWith('.ts')
			? intputFile.slice(0, -'.ts'.length) + '.d.ts'
			: intputFile.endsWith('.tsx')
				? intputFile.slice(0, -'.tsx'.length) + '.d.ts'
				: intputFile + '.d.ts';
		it(`Input: ${shortenPath(intputFile)}, Output: ${shortenPath(expectedOutputFile)}`, () => {
			let outputText: string | undefined;
			const sourceFile = program.getSourceFile(intputFile);
			program.emit(
				sourceFile,
				(outputFile, text) => {
					expect(outputFile.replace(windowsPathReg, '/')).toBe(expectedOutputFile.replace(windowsPathReg, '/'));
					text = removeEmitGlobalTypes(text);
					outputText = text;
				},
				undefined,
				true,
			);
			expect(outputText ? normalizeNewline(outputText) : undefined).toMatchSnapshot();
		});
	}
});

function readFilesRecursive(dir: string) {
	const result: string[] = [];

	for (const file of fs.readdirSync(dir)) {
		if (file === 'tsconfig.json') {
			continue;
		}
		const filepath = path.join(dir, file);
		const stat = fs.statSync(filepath);
		if (stat.isDirectory()) {
			result.push(...readFilesRecursive(filepath));
		} else {
			result.push(filepath);
		}
	}

	return result;
}

function shortenPath(path: string): string {
	path = normalizePath(path);
	const segments = path.split('/');
	return segments.slice(-2).join('/');
}
