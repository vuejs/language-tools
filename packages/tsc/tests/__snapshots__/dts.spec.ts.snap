// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue-tsc-dts > Input: class-slots/component.vue, Output: class-slots/component.vue.d.ts 1`] = `
"import { VNode } from 'vue';
declare const _default: new () => {
    $slots: {
        default(_: {
            num: number;
        }): VNode[];
        foo(_: {
            str: string;
        }): VNode[];
    };
};
export default _default;
"
`;

exports[`vue-tsc-dts > Input: empty-component/component.vue, Output: empty-component/component.vue.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: empty-component/custom-extension-component.cext, Output: empty-component/custom-extension-component.cext.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: events/component-class.vue, Output: events/component-class.vue.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {
    foo: (value: string) => void;
}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>> & {
    onFoo?: (value: string) => any;
}, {}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: events/component-generic.vue, Output: events/component-generic.vue.d.ts 1`] = `
"declare const _default: <T>(__VLS_props: {
    onFoo?: (value: string) => any;
} & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, __VLS_ctx?: {
    slots: {};
    attrs: any;
    emit: (evt: "foo", value: string) => void;
}, __VLS_expose?: (exposed: import('vue').ShallowUnwrapRef<{}>) => void, __VLS_setup?: Promise<{
    props: {
        onFoo?: (value: string) => any;
    } & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps;
    expose(exposed: import('vue').ShallowUnwrapRef<{}>): void;
    attrs: any;
    slots: {};
    emit: (evt: "foo", value: string) => void;
}>) => import("vue").VNode<import("vue").RendererNode, import("vue").RendererElement, {
    [key: string]: any;
}> & {
    __ctx?: {
        props: {
            onFoo?: (value: string) => any;
        } & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps;
        expose(exposed: import('vue').ShallowUnwrapRef<{}>): void;
        attrs: any;
        slots: {};
        emit: (evt: "foo", value: string) => void;
    };
};
export default _default;
type __VLS_Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
"
`;

exports[`vue-tsc-dts > Input: generic/component.vue, Output: generic/component.vue.d.ts 1`] = `
"declare const _default: <T>(__VLS_props: {
    onBar?: (data: number) => any;
    foo: number;
} & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, __VLS_ctx?: {
    slots: Readonly<{
        default?(data: {
            foo: number;
        }): any;
    }> & {
        default?(data: {
            foo: number;
        }): any;
    };
    attrs: any;
    emit: (e: 'bar', data: number) => void;
}, __VLS_expose?: (exposed: import("vue").ShallowUnwrapRef<{
    baz: number;
}>) => void, __VLS_setup?: Promise<{
    props: {
        onBar?: (data: number) => any;
        foo: number;
    } & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps;
    expose(exposed: import("vue").ShallowUnwrapRef<{
        baz: number;
    }>): void;
    attrs: any;
    slots: Readonly<{
        default?(data: {
            foo: number;
        }): any;
    }> & {
        default?(data: {
            foo: number;
        }): any;
    };
    emit: (e: 'bar', data: number) => void;
}>) => import("vue").VNode<import("vue").RendererNode, import("vue").RendererElement, {
    [key: string]: any;
}> & {
    __ctx?: {
        props: {
            onBar?: (data: number) => any;
            foo: number;
        } & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps;
        expose(exposed: import("vue").ShallowUnwrapRef<{
            baz: number;
        }>): void;
        attrs: any;
        slots: Readonly<{
            default?(data: {
                foo: number;
            }): any;
        }> & {
            default?(data: {
                foo: number;
            }): any;
        };
        emit: (e: 'bar', data: number) => void;
    };
};
export default _default;
type __VLS_Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
"
`;

exports[`vue-tsc-dts > Input: generic/custom-extension-component.cext, Output: generic/custom-extension-component.cext.d.ts 1`] = `
"declare const _default: <T>(__VLS_props: {
    onBar?: (data: number) => any;
    foo: number;
} & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, __VLS_ctx?: {
    slots: Readonly<{
        default?(data: {
            foo: number;
        }): any;
    }> & {
        default?(data: {
            foo: number;
        }): any;
    };
    attrs: any;
    emit: (e: 'bar', data: number) => void;
}, __VLS_expose?: (exposed: import("vue").ShallowUnwrapRef<{
    baz: number;
}>) => void, __VLS_setup?: Promise<{
    props: {
        onBar?: (data: number) => any;
        foo: number;
    } & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps;
    expose(exposed: import("vue").ShallowUnwrapRef<{
        baz: number;
    }>): void;
    attrs: any;
    slots: Readonly<{
        default?(data: {
            foo: number;
        }): any;
    }> & {
        default?(data: {
            foo: number;
        }): any;
    };
    emit: (e: 'bar', data: number) => void;
}>) => import("vue").VNode<import("vue").RendererNode, import("vue").RendererElement, {
    [key: string]: any;
}> & {
    __ctx?: {
        props: {
            onBar?: (data: number) => any;
            foo: number;
        } & import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps;
        expose(exposed: import("vue").ShallowUnwrapRef<{
            baz: number;
        }>): void;
        attrs: any;
        slots: Readonly<{
            default?(data: {
                foo: number;
            }): any;
        }> & {
            default?(data: {
                foo: number;
            }): any;
        };
        emit: (e: 'bar', data: number) => void;
    };
};
export default _default;
type __VLS_Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
"
`;

exports[`vue-tsc-dts > Input: non-component/component.ts, Output: non-component/component.d.ts 1`] = `
"declare const _default: {};
export default _default;
"
`;

exports[`vue-tsc-dts > Input: options-api/component.ts, Output: options-api/component.d.ts 1`] = `
"interface SubmitPayload {
    /**
     * email of user
     */
    email: string;
    /**
     * password of same user
     */
    password: string;
}
declare const _default: import("vue").DefineComponent<{
    /**
     * Default number
     */
    numberDefault: {
        type: NumberConstructor;
        default: number;
    };
    /**
     * Default function Object
     */
    objectDefault: {
        type: ObjectConstructor;
        default: () => {
            foo: string;
        };
    };
    /**
     * Default function Array
     */
    arrayDefault: {
        type: ArrayConstructor;
        default: () => number[];
    };
    /**
     * Default function more complex
     */
    complexDefault: {
        type: ArrayConstructor;
        default: (props: any) => any[];
    };
}, unknown, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {
    submit: ({ email, password }: SubmitPayload) => boolean;
}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{
    /**
     * Default number
     */
    numberDefault: {
        type: NumberConstructor;
        default: number;
    };
    /**
     * Default function Object
     */
    objectDefault: {
        type: ObjectConstructor;
        default: () => {
            foo: string;
        };
    };
    /**
     * Default function Array
     */
    arrayDefault: {
        type: ArrayConstructor;
        default: () => number[];
    };
    /**
     * Default function more complex
     */
    complexDefault: {
        type: ArrayConstructor;
        default: (props: any) => any[];
    };
}>> & {
    onSubmit?: (args_0: SubmitPayload) => any;
}, {
    numberDefault: number;
    objectDefault: Record<string, any>;
    arrayDefault: unknown[];
    complexDefault: unknown[];
}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: reference-type-events/component.vue, Output: reference-type-events/component.vue.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {
    foo: (data?: {
        foo: string;
    }) => void;
    bar: (value: {
        arg1: number;
        arg2?: any;
    }) => void;
    baz: () => void;
}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>> & {
    onFoo?: (data?: {
        foo: string;
    }) => any;
    onBar?: (value: {
        arg1: number;
        arg2?: any;
    }) => any;
    onBaz?: () => any;
}, {}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: reference-type-events/my-events.ts, Output: reference-type-events/my-events.d.ts 1`] = `
"/**
 * Emitted when the foo property is changed.
 */
type MyEventsFoo = 'foo';
export interface MyEvents {
    (event: MyEventsFoo, data?: {
        foo: string;
    }): void;
    (event: 'bar', value: {
        arg1: number;
        arg2?: any;
    }): void;
    (e: 'baz'): void;
}
export {};
"
`;

exports[`vue-tsc-dts > Input: reference-type-exposed/component.vue, Output: reference-type-exposed/component.vue.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{}, {
    /**
     * a counter string
     */
    counter: import("vue").Ref<string>;
}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: reference-type-model/component.vue, Output: reference-type-model/component.vue.d.ts 1`] = `
"type __VLS_PublicProps = {
    "foo"?: number;
    "bar"?: string[];
    "qux"?: string;
    quxModifiers?: Record<'lazy' | 'trim', true>;
};
declare const _default: import("vue").DefineComponent<__VLS_TypePropsToOption<__VLS_PublicProps>, {}, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {
    "update:foo": (foo: number) => void;
    "update:bar": (bar: string[]) => void;
    "update:qux": (qux: string) => void;
}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<__VLS_TypePropsToOption<__VLS_PublicProps>>> & {
    "onUpdate:foo"?: (foo: number) => any;
    "onUpdate:bar"?: (bar: string[]) => any;
    "onUpdate:qux"?: (qux: string) => any;
}, {}, {}>;
export default _default;
type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;
type __VLS_TypePropsToOption<T> = {
    [K in keyof T]-?: {} extends Pick<T, K> ? {
        type: import('vue').PropType<__VLS_NonUndefinedable<T[K]>>;
    } : {
        type: import('vue').PropType<T[K]>;
        required: true;
    };
};
"
`;

exports[`vue-tsc-dts > Input: reference-type-props/component.vue, Output: reference-type-props/component.vue.d.ts 1`] = `
"import { MyProps } from './my-props';
declare const _default: import("vue").DefineComponent<__VLS_WithDefaults<__VLS_TypePropsToOption<MyProps>, {
    bar: number;
    baz: () => string[];
}>, {}, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<__VLS_WithDefaults<__VLS_TypePropsToOption<MyProps>, {
    bar: number;
    baz: () => string[];
}>>>, {
    bar: number;
    baz: string[];
}, {}>;
export default _default;
type __VLS_WithDefaults<P, D> = {
    [K in keyof Pick<P, keyof P>]: K extends keyof D ? __VLS_Prettify<P[K] & {
        default: D[K];
    }> : P[K];
};
type __VLS_Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;
type __VLS_TypePropsToOption<T> = {
    [K in keyof T]-?: {} extends Pick<T, K> ? {
        type: import('vue').PropType<__VLS_NonUndefinedable<T[K]>>;
    } : {
        type: import('vue').PropType<T[K]>;
        required: true;
    };
};
"
`;

exports[`vue-tsc-dts > Input: reference-type-props/component-js.vue, Output: reference-type-props/component-js.vue.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{
    foo: {
        type: StringConstructor;
        required: true;
    };
    bar: {
        type: StringConstructor;
        default: string;
    };
    baz: {
        type: StringConstructor;
    };
    xfoo: {
        readonly type: StringConstructor;
        readonly required: true;
    };
    xbar: {
        readonly type: StringConstructor;
        readonly value: "";
    };
    xbaz: {
        readonly type: StringConstructor;
    };
}, unknown, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{
    foo: {
        type: StringConstructor;
        required: true;
    };
    bar: {
        type: StringConstructor;
        default: string;
    };
    baz: {
        type: StringConstructor;
    };
    xfoo: {
        readonly type: StringConstructor;
        readonly required: true;
    };
    xbar: {
        readonly type: StringConstructor;
        readonly value: "";
    };
    xbaz: {
        readonly type: StringConstructor;
    };
}>>, {
    bar: string;
}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: reference-type-props/component-js-setup.vue, Output: reference-type-props/component-js-setup.vue.d.ts 1`] = `
"declare const _default: import("vue").DefineComponent<{
    foo: {
        type: StringConstructor;
        required: true;
    };
    bar: {
        type: StringConstructor;
        default: string;
    };
    baz: {
        type: StringConstructor;
    };
    xfoo: {
        readonly type: StringConstructor;
        readonly required: true;
    };
    xbar: {
        readonly type: StringConstructor;
        readonly value: "";
    };
    xbaz: {
        readonly type: StringConstructor;
    };
    /**
     * The hello property.
     *
     * @since v1.0.0
     */
    hello: {
        type: StringConstructor;
        default: string;
    };
    numberOrStringProp: {
        type: (NumberConstructor | StringConstructor)[];
        default: number;
    };
    arrayProps: {
        type: ArrayConstructor;
        default: () => number[];
    };
}, {}, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{
    foo: {
        type: StringConstructor;
        required: true;
    };
    bar: {
        type: StringConstructor;
        default: string;
    };
    baz: {
        type: StringConstructor;
    };
    xfoo: {
        readonly type: StringConstructor;
        readonly required: true;
    };
    xbar: {
        readonly type: StringConstructor;
        readonly value: "";
    };
    xbaz: {
        readonly type: StringConstructor;
    };
    /**
     * The hello property.
     *
     * @since v1.0.0
     */
    hello: {
        type: StringConstructor;
        default: string;
    };
    numberOrStringProp: {
        type: (NumberConstructor | StringConstructor)[];
        default: number;
    };
    arrayProps: {
        type: ArrayConstructor;
        default: () => number[];
    };
}>>, {
    bar: string;
    hello: string;
    numberOrStringProp: string | number;
    arrayProps: unknown[];
}, {}>;
export default _default;
"
`;

exports[`vue-tsc-dts > Input: reference-type-props/my-props.ts, Output: reference-type-props/my-props.d.ts 1`] = `
"export interface MyNestedProps {
    /**
     * nested prop documentation
     */
    nestedProp: string;
}
export interface MyIgnoredNestedProps {
    nestedProp: string;
}
export interface MyNestedRecursiveProps {
    recursive: MyNestedRecursiveProps;
}
declare enum MyEnum {
    Small = 0,
    Medium = 1,
    Large = 2
}
declare const categories: readonly ["Uncategorized", "Content", "Interaction", "Display", "Forms", "Addons"];
type MyCategories = typeof categories[number];
export interface MyProps {
    /**
   * string foo
   *
   * @default "rounded"
   * @since v1.0.0
   * @see https://vuejs.org/
   * @example
   * \`\`\`vue
   * <template>
   *   <component foo="straight" />
   * </template>
   * \`\`\`
     */
    foo: string;
    /**
     * optional number bar
     */
    bar?: number;
    /**
     * string array baz
     */
    baz?: string[];
    /**
     * required union type
     */
    union: string | number;
    /**
     * optional union type
     */
    unionOptional?: string | number;
    /**
     * required nested object
     */
    nested: MyNestedProps;
    /**
     * required nested object with intersection
     */
    nestedIntersection: MyNestedProps & {
        /**
         * required additional property
         */
        additionalProp: string;
    };
    /**
     * optional nested object
     */
    nestedOptional?: MyNestedProps | MyIgnoredNestedProps;
    /**
     * required array object
     */
    array: MyNestedProps[];
    /**
     * optional array object
     */
    arrayOptional?: MyNestedProps[];
    /**
     * enum value
     */
    enumValue: MyEnum;
    /**
     * literal type alias that require context
     */
    literalFromContext: MyCategories;
    inlined: {
        foo: string;
    };
    recursive: MyNestedRecursiveProps;
}
export declare const StringRequired: {
    readonly type: StringConstructor;
    readonly required: true;
};
export declare const StringEmpty: {
    readonly type: StringConstructor;
    readonly value: "";
};
export declare const StringUndefined: {
    readonly type: StringConstructor;
};
export {};
"
`;

exports[`vue-tsc-dts > Input: template-slots/component.vue, Output: template-slots/component.vue.d.ts 1`] = `
"declare const _default: __VLS_WithTemplateSlots<import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>, {
    "no-bind"?(_: {}): any;
    default?(_: {
        num: number;
    }): any;
    "named-slot"?(_: {
        str: string;
    }): any;
    vbind?(_: {
        num: number;
        str: string;
    }): any;
}>;
export default _default;
type __VLS_WithTemplateSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
"
`;

exports[`vue-tsc-dts > Input: template-slots/component-define-slots.vue, Output: template-slots/component-define-slots.vue.d.ts 1`] = `
"import { VNode } from 'vue';
declare const _default: __VLS_WithTemplateSlots<import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>, Readonly<{
    default: (props: {
        num: number;
    }) => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
    'named-slot': (props: {
        str: string;
    }) => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
    vbind: (props: {
        num: number;
        str: string;
    }) => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
    'no-bind': () => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
}> & {
    default: (props: {
        num: number;
    }) => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
    'named-slot': (props: {
        str: string;
    }) => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
    vbind: (props: {
        num: number;
        str: string;
    }) => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
    'no-bind': () => VNode<import("vue").RendererNode, import("vue").RendererElement, {
        [key: string]: any;
    }>[];
}>;
export default _default;
type __VLS_WithTemplateSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
"
`;

exports[`vue-tsc-dts > Input: template-slots/component-destructuring.vue, Output: template-slots/component-destructuring.vue.d.ts 1`] = `
"declare const _default: __VLS_WithTemplateSlots<import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>, Readonly<{
    bottom: (props: {
        num: number;
    }) => any[];
}> & {
    bottom: (props: {
        num: number;
    }) => any[];
}>;
export default _default;
type __VLS_WithTemplateSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
"
`;

exports[`vue-tsc-dts > Input: template-slots/component-no-script.vue, Output: template-slots/component-no-script.vue.d.ts 1`] = `
"declare const _default: __VLS_WithTemplateSlots<import("vue").DefineComponent<{}, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{}>>, {}, {}>, {
    "no-bind"?(_: {}): any;
    default?(_: {
        num: number;
    }): any;
    "named-slot"?(_: {
        str: string;
    }): any;
    vbind?(_: {
        num: number;
        str: string;
    }): any;
}>;
export default _default;
type __VLS_WithTemplateSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
"
`;

exports[`vue-tsc-dts > Input: ts-component/PropDefinitions.ts, Output: ts-component/PropDefinitions.d.ts 1`] = `
"export interface MyProps {
    /**
     * string foo
     */
    foo: string;
    /**
     * optional number bar
     */
    bar?: number;
}
"
`;

exports[`vue-tsc-dts > Input: ts-component/component.ts, Output: ts-component/component.d.ts 1`] = `undefined`;

exports[`vue-tsc-dts > Input: ts-component/component.tsx, Output: ts-component/component.d.ts 1`] = `undefined`;

exports[`vue-tsc-dts > Input: ts-named-export/component.ts, Output: ts-named-export/component.d.ts 1`] = `
"export declare const Foo: import("vue").DefineSetupFnComponent<{
    foo: string;
}, {}, {}, {
    foo: string;
} & {}, import("vue").PublicProps>;
export declare const Bar: import("vue").DefineSetupFnComponent<{
    bar?: number;
}, {}, {}, {
    bar?: number;
} & {}, import("vue").PublicProps>;
"
`;
